### DISCUSSION TITLE
Basic AST for templating engines

### DISCUSSION THREAD

**Discussion by binyamin** [syntax-tree/unist#59] [Ideas] [answered]
I had this idea, that unist could (_somehow_) emulate libraries like nunjucks, liquid, and handlebars. This way, the document would be only be parsed and written once.

I don't really know where this would fit into the unified ecosystem. I feel like it would need a new AST, with plugins for other ecosystems (eg. remark, rehype). I did have a look at [unifiedjs/collective](https://github.com/unifiedjs/collective/), so I know I would need to start a motion to make this official.

What logic needs to be defined, in order for this to become a reality?

**Comment by wooorm** [comment, marked as answer]
Hey!

I think where it makes sense depends some more on what you’re thinking of. Are you thinking of a particular language? Multiple? What are the “units” in the tree? Interpolation, calls, strings, what else? What does an AST for this solve? How do you mean it would only be parsed once and written once (I think I get it, but just to be sure)?

I don’t think we need to prematurely figure out how it all should be organized. Projects working with unified/unist can exist fine on their own outside of “our” maintenance (a small example is: https://github.com/wooorm/dioscuri)

**Reply by binyamin** [reply]
I'm thinking more of a generic AST, which could be used to implement one of many templating engines. As for the units, I would compare some popular syntaxes, synthesize the results, and define an AST accordingly. It would incorporate features like variables, filters assignment tags, and logic blocks like for and if/else.

Personally, I think the AST would be super-useful for static-site generators. It would help with speed, memory consumption - all the good stuff.
![A diagram illustrates how, with this generic AST, the document would be parsed once (by unified) instead of twice (by the templating engine as well).](https://user-images.githubusercontent.com/39805353/148608191-849f45e6-de7b-4a4e-8090-fcbcf94d18ba.png)

**Reply by wooorm** [reply]
Thanks, yeah, I can see the value! Though maybe the “in theory” works less well than hoped.
I also don’t really use template languages (though I have some previous experience), so I can’t be of much help.

**Reply by ChristianMurphy** [reply]
I also see the potential in this, though something worth noting on this.

Unified itself isn't a language parser; remark, rehype, retext, redot, etc do the language parsing.
For parsing to happen in a single pass, the language in question would need to be aware of the templating language so that it can parse it. (for example a similar past discussion on remark https://github.com/remarkjs/remark/discussions/922)

It sounds like this would likely need to extend remark (similar to mdx) or extend rehype. Unless you are planning for it to be completely independent of any of those, and just do templating.
Which would also be fine, but could get back to double parsing, if also used with rehype or remark.

**Comment by linonetwo** [comment]
I've created this https://github.com/linonetwo/template-based-generator-template using mdast utils to parse md based template, and fill slots using variables inside JSON, and serialize using nlcst utils.
